### PR TITLE
Cached trie node estimation

### PR DESCRIPTION
**Pull request by jakmeier**
Add an estimation for the new gas cost introduced in #6588.

The cost is called `read_cached_trie_node` and is charged whenever a trie node is read from the in-memory cache instead of from the database. This is a numerically small cost but one that is charged often.

The exact cost is hard to measure because it almost exclusively depends on memory latency, which is different from machine to machine. Plus, the state of CPU caches drastically influences memory latency as well.

On a high-level, the estimation overcomes the challenges by setting up an environment that
(1) can run an isolated workload that is equivalent to reading a node from the trie node cache without any overhead,
(2) allows to tune assumptions around memory latency.

Then the estimation measures a sample distribution of the workload for a pessimistic memory latency assumption.

Using the `--debug` flag of the estimator will also measure the distribution for other memory assumptions and print a table with the comparison.

### CODE REVIEW

**Comment by jakmeier** [comment]
This should be ready now for review and merging into master. With the `--debug` flag we can still run different assumptions for memory accesses. But for regular estimation, only one case is tested and it's the 90th percentile of 200 samples that is used as output. 

All results are still a bit volatile, unfortunately. We could add more iterations but I don't want to slow it down too much.

At least, running with `--iters 3` and `--warmup-iters 1` currently gives me fairly consistent estimation outputs on gcloud, in the range of 1450-1550Mgas.

File: runtime/runtime-params-estimator/Cargo.toml
Comment (Longarithm) [inline runtime/runtime-params-estimator/Cargo.toml:63]: Could you remove all `protocol_feature_chunk_nodes_cache` occurrences? I guess it got here from the old chunk cache PR, the feature is stable now.

Comment (jakmeier) [reply runtime/runtime-params-estimator/Cargo.toml]: Should be fixed now

**Comment by jakmeier** [comment]
This new estimation, the results of which are already used in the last released, still need a review from a code owner.

@matklad or @nagisa could maybe one of you take a look at this? 

Besides just being an estimation for `read_cached_trie_node` cost, I think this also shows a possible direction for other estimations that are highly dependent on memory latency. That is, we wan to assume "bad" memory latency but we cannot really commit to worst-case. Measuring a range of cases, determining a realistic case, and taking the 90th percentile of the samples measured in the realistic case is the approach I use here and would suggest to use more in the future.

File: runtime/runtime-params-estimator/src/trie.rs
Comment (matklad) [inline runtime/runtime-params-estimator/src/trie.rs:409]:
```suggestion
                _ => unreachable!(),
```

File: runtime/runtime-params-estimator/src/trie.rs
Comment (matklad) [inline runtime/runtime-params-estimator/src/trie.rs:309]: Nice code layout :+1: !

File: runtime/runtime-params-estimator/src/trie.rs
Comment (matklad) [inline runtime/runtime-params-estimator/src/trie.rs:375]: I wonder if we should just have 

```
pub fn spoil_l3_cache() {}
```

utility function?

Or would it actually require `L3Spoiler` struct to cache the random data? 

Not an action item, mostly just thoughts!

File: core/store/src/trie/mod.rs
Comment (matklad) [inline core/store/src/trie/mod.rs:267]: It's unfortunate that we have to make a bunch of stuff public, but I don't really see a way around this. 

Though, *maaaybe* if we can add something a touch more encapsulated here?

like 

```rust
pub mod estimator {
   pub fn encode_extension_node(key: &[u8]) -> Vec<u8>
}
```

The problem with making stuff `pub` is that it's not clear that it is pub for this one very specific use-case, rather "generally pub".

File: runtime/runtime-params-estimator/src/trie.rs
Comment (matklad) [inline runtime/runtime-params-estimator/src/trie.rs:349]: Wait, what is happening here? I am hopelessly confused here -- to me, it seems that this block should be roughly equivalent to just `tb.random_vec(value_len)`?

Like, we encode a raw trie node, but then use this thing as a *value* we just insert? Ie, what we get in a database is a trie node whose payload is an encoded trie node, no?

**Review by matklad — COMMENTED** [review]
This all sounds very reasonable!

And :+1: on the overall estimation strategy! I think that the fact that we know very specific parameters which we can very to go from worst-case to best-case gives us more evidence that we actually *understand* what the worst case is!

However, I am confused about the RawTrieNode stuff we do in `read_node_from_chunk_cache_ext`, could you explain what's going on there? Either I completely misunderstand the code, or there's some kind of double encoding happening.

File: runtime/runtime-params-estimator/src/trie.rs
Comment (jakmeier) [reply runtime/runtime-params-estimator/src/trie.rs:349]: Oof, looking back at this, I am confused myself. :sweat_smile: At the very least, it needs some comments.

This evolved as @Longarithm and I were trying different estimation strategies. I actually didn't write the difficult part of the code but I hope understood it right. Let me explain.

Yes, it does store an encoded trie node as the value of a node. And it is confusing, yet intentional. 

The goal here is two-fold.
1. Use the trie cache directly (i.e. without the trie) but keep the encoding and decoding logic. Basically, emulate the normal behaviour to lookup a trie node, which is this:
```rust
fn retrieve_node(&self, hash: &CryptoHash) -> Result<TrieNodeWithSize, StorageError> {
    if *hash == Trie::empty_root() {
        return Ok(TrieNodeWithSize::empty());
    }
    let bytes = self.storage.retrieve_raw_bytes(hash)?;
    match RawTrieNodeWithSize::decode(&bytes) {
        Ok(value) => Ok(TrieNodeWithSize::from_raw(value)),
        Err(_) => Err(StorageError::StorageInconsistentState(format!(
            "Failed to decode node {}",
            hash
        ))),
    }
}
```
2. Read a value that represents the worst-case for cached node lookup. And the worst-case is, an extension trie node with a 2kB long key in it. Why? Because reading most nodes is dominated by other costs, such as `storage_read_value_byte` for leaf nodes and `storage_read_key_byte` for branches and extension we just iterate over. However, when we read an extension node form the cache because we look up the value stored _at_ the extension point, in that case we read the key extension of the value _below_ the node as well. And this key extension can be potentially 2k bytes long and is not part of the gas equation. So that is the worst-case we care about in this estimation.

Generating many different extension nodes inside the trie is a bit tricky, since they interact with each other and we don't want keys to become shorter than 2kB. So it would be nice to have a more direct way to just insert trie nodes that look like our worst case, so we read those into the cache.

The tricky part on the implementation side is, the interface `trait TrieStorage` only reads values. Encoding and storing of values is done on layers above that. Thus, the trick is to insert an encoded trie node as a value through normal block processing. When inserting `("key", "value")` like that, a bunch of actual trie nodes for `"key"` are inserted to RocksDB, all of which we ignore completely. All we need is the raw `"value"`, which is also inserted by its hash as the key.

Now, since we know the hash, we can obtain our raw node value through the trie storage interface and decode it like a real trie node. 

I admit, this is rather insane... But the thing we want to measure it too small that we could allow extra overhead. Any kind of subtraction between two costs we tried was highly unstable.

File: runtime/runtime-params-estimator/src/transaction_builder.rs
Comment (jakmeier) [reply runtime/runtime-params-estimator/src/transaction_builder.rs:78]: Nope, let me merge the two

File: runtime/runtime-params-estimator/src/trie.rs
Comment (jakmeier) [reply runtime/runtime-params-estimator/src/trie.rs:379]: Yes, I think that would work as utility function! I will consider refactoring, next time I need to spoil L3 in my code.

But, I would use it somewhat sparingly. It can slow down estimations quite a bit. Probably only worth it for costs that are memory latency bound.

File: core/store/src/trie/mod.rs
Comment (jakmeier) [reply core/store/src/trie/mod.rs:267]: I agree. And I like your suggestion to, let me implement that.

File: runtime/runtime-params-estimator/src/trie.rs
Comment (matklad) [reply runtime/runtime-params-estimator/src/trie.rs:353]: That's smart! But would definitely benefit from a comment :-)

File: core/store/src/trie/mod.rs
Comment (jakmeier) [reply core/store/src/trie/mod.rs]: Done, now the types and the "real" encoding / decoding functions are back to being private, as they should be.

**Review by matklad — APPROVED** [review]

**Comment by mina86** [comment]
Could you put in PR description what this is actually doing on high level? It’s a bit hard to understand for someone unfamiliar with the code.

**Comment by jakmeier** [comment]
> Could you put in PR description what this is actually doing on high level? It’s a bit hard to understand for someone unfamiliar with the code.

A bit of an understatement if you ask me. :sweat_smile: 

I now added a high-level description to the PR. I think for people who are not familiar with both the estimator code or the trie code, it will still be rather difficult to understand. But I don't think I can do much to change that...

@mina86 , since we already have a code owner review and the code only concerns estimations, probably you don't need to spend the time to thoroughly understand all the details. But does the new comment at least make the high-level intention of the PR clear enough?

File: runtime/runtime-params-estimator/src/utils.rs
Comment (mina86) [inline runtime/runtime-params-estimator/src/utils.rs:265]: `percentiles.into_iter().map(|p| (p * costs.len() as f32).ceil() as usize - 1).map(|idx| costs[idx].clone()).collect()`?

File: runtime/runtime-params-estimator/src/utils.rs
Comment (mina86) [inline runtime/runtime-params-estimator/src/utils.rs:301]: You could change the third argument to `&[u64]` and then you wouldn’t need to create the vectors.  It is a test so we don’t really care about performance but to me it feels cleaner regardless.

File: runtime/runtime-params-estimator/src/utils.rs
Comment (mina86) [inline runtime/runtime-params-estimator/src/utils.rs:258]: Alternatively `-> impl Iterator<Item=GasCost>` since in most cases we’re later only iterating over the vector.

File: runtime/runtime-params-estimator/src/trie.rs
Comment (mina86) [inline runtime/runtime-params-estimator/src/trie.rs:26]: `let value = b"0";` and you won’t need `.as_bytes()`.  Similarly below with `[b"1", b"2", b"3"]`.  Same applies to read_node_from_db.

File: runtime/runtime-params-estimator/src/trie.rs
Comment (mina86) [inline runtime/runtime-params-estimator/src/trie.rs:178]: It looks to me like this should be put in its own function.  We’re doing the same thing over and over only with the parameters changing.  Could be local function or even local closure so you wouldn’t need to pass percentiles_of_interest and testbed as arguments.

File: runtime/runtime-params-estimator/src/trie.rs
Comment (mina86) [inline runtime/runtime-params-estimator/src/trie.rs:22]: `"j".repeat(final_key_length)` like in the other function?

File: runtime/runtime-params-estimator/src/trie.rs
Comment (mina86) [inline runtime/runtime-params-estimator/src/trie.rs:25]: No need for `.as_str()`. You can even put indexing at the end: `let key = &key.as_bytes()[..key_len];`. Less syntax feels cleaner.  Same applies to read_node_from_db.

**Review by mina86 — APPROVED** [review]
LGTM, though honestly I’m still not sure if I fully understand the code. I’m assuming Aleksey does.s ;) Just a few stylistic suggestions.

**Comment by jakmeier** [comment]
Thanks @mina86 for the valuable inputs, I incorporated all changes you suggested. Removes 78 LOC for doing the same thing. And the code is now more readable than before. (Although, it still complex in what it does.)

File: runtime/runtime-params-estimator/src/trie.rs
Comment (jakmeier) [reply runtime/runtime-params-estimator/src/trie.rs]: Yeah I've put it in a closure, looks much better now IMO

**Pull request merged**
